pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -627,6 +628,316 @@ func (og *operationGenerator) GenerateUnmountDeviceFunc(
+		markVolUnmountedErr := actualStateOfWorld.MarkVolumeAsUnmounted(
+			volumeToUnmount.PodName, volumeToUnmount.VolumeName)
+		if markVolUnmountedErr != nil {
+			// On failure, just log and exit
+			glog.Errorf(volumeToUnmount.GenerateErrorDetailed("UnmapVolume.MarkVolumeAsUnmounted failed", markVolUnmountedErr).Error())
+		}
+
+		return nil
+	}, blockVolumePlugin.GetPluginName(), nil
+}
+
+func (og *operationGenerator) GenerateUnmapDeviceFunc(
+	deviceToDetach AttachedVolume,
+	actualStateOfWorld ActualStateOfWorldMounterUpdater,
+	mounter mount.Interface) (func() error, string, error) {
+
+	// Get block volume mapper plugin
+	var blockVolumeMapper volume.BlockVolumeMapper
+	blockVolumePlugin, err :=
+		og.volumePluginMgr.FindMapperPluginBySpec(deviceToDetach.VolumeSpec)
+	if err != nil {
+		return nil, "", deviceToDetach.GenerateErrorDetailed("UnmapDevice.FindMapperPluginBySpec failed", err)
+	}
+	if blockVolumePlugin == nil {
+		return nil, "", deviceToDetach.GenerateErrorDetailed("UnmapDevice.FindMapperPluginBySpec failed to find BlockVolumeMapper plugin. Volume plugin is nil.", nil)
+	}
+	blockVolumeMapper, newMapperErr := blockVolumePlugin.NewBlockVolumeMapper(
+		deviceToDetach.VolumeSpec,
+		nil, /* Pod */
+		volume.VolumeOptions{})
+	if newMapperErr != nil {
+		return nil, "", deviceToDetach.GenerateErrorDetailed("UnmapDevice.NewBlockVolumeMapper initialization failed", newMapperErr)
+	}
+
+	blockVolumeUnmapper, newUnmapperErr := blockVolumePlugin.NewBlockVolumeUnmapper(
+		string(deviceToDetach.VolumeName),
+		"" /* podUID */)
+	if newUnmapperErr != nil {
+		return nil, blockVolumePlugin.GetPluginName(), deviceToDetach.GenerateErrorDetailed("UnmapDevice.NewUnmapper failed", newUnmapperErr)
+	}
+
+	return func() error {
+		globalMapPath, err :=
+			blockVolumeMapper.GetGlobalMapPath(deviceToDetach.VolumeSpec)
+		if err != nil {
+			// On failure, return error. Caller will log and retry.
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice.GetGlobalMapPath failed", err)
+		}
+		refs, err := util.GetDeviceSymlinkRefs(deviceToDetach.DevicePath, globalMapPath)
+		if err != nil {
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice.GetDeviceSymlinkRefs check failed", err)
+		}
+		if len(refs) > 0 {
+			err = fmt.Errorf("The device %q is still referenced from other Pods %v", globalMapPath, refs)
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice failed", err)
+		}
+		// The globalMapPath directory is empty. Remove the directory
+		removeMapPathErr := util.RemoveMapPath(globalMapPath)
+		if removeMapPathErr != nil {
+			// On failure, return error. Caller will log and retry.
+			return deviceToDetach.GenerateErrorDetailed("UnmapDevice failed", removeMapPathErr)
+		}
+
+		// The block volume is not referenced from Pods. Release file descriptor lock.
+		glog.V(5).Infof("UnmapDevice: deviceToDetach.DevicePath: %v", deviceToDetach.DevicePath)
+		exec := og.volumePluginMgr.Host.GetExec(blockVolumePlugin.GetPluginName())
```

Comment:
This requires that losetup is in all containers with mount utilities. You probably want to execute the command on the host (or in the same container where is kubelet), hece use directly os/exec. And you probably don't need to pass the exec interface around, GetLoopDevice/RemoveLoopDevice et al can call exec directly.

This applies also to AttachFileDevice above.